### ISSUE TITLE
Double inheritance bug not reported

### ISSUE THREAD

**Issue by socketpair** [PyCQA/pylint#6992]
### Bug description

class A:
   def close(self):
      print('A')

class B:
   def close(self):
      print('B')

class C(A, B):
   def close(self):
      print('C')
      super().close()

C().stop()

### Command used

```shell
pylint a.py
```

### Pylint output

```shell
(nothing)
```

### Expected behavior

"B.close() is not called" message

### Pylint version

```shell
latest
```

**Comment by DanielNoord** [comment]
@socketpair Please use the issue template to report bugs. And please don't remove the template after you have just used it, there is reason open source maintainers enforce such templates.

I wonder how many people use this pattern without thinking it is a bug. We have ``super-init-not-called``. Should we extend that cover all methods?

**Comment by socketpair** [comment]
@DanielNoord 
> Should we extend that cover all methods?
For sure, we should. It would not be a bug if one of the classes  (A or B) did not have the `close()` method.

**Comment by DanielNoord** [comment]
What about:
```python
class A:
   def close(self):
      print('A')

class C(A):
   def close(self):
      print('C')
```

I don't think that should raise right? So it would only cover methods that have a `super()` call within them that does not cover all parents? That's different for `__init__` which should always raise whenever we detect that `super()` isn't called.

But then again, the method works without calling the parent, whereas the class will be broken if the constructor isn't called.

**Comment by socketpair** [comment]
@DanielNoord

1. I did not add constructors to all the classes because it is irrelevant to the problem. For a complete example, yes, we should add constructor to C and call both base constructors from it.
2. Regarding your case, in most real cases it IS a bug. It's definitely not a bug if:

```python
class A:
   def close(self):
      raise Not ImplementedError

class C(A):
   def close(self):
      print('C')
```

or
 
```python
class A:
   @abstractmethod
   def close(self):
      pass

class C(A):
   def close(self):
      print('C')
```

For the cases when one really want to redefine logic (as in case you provided), yes, he really should add something like `#  pylint: disable=base-method-not-called`.

**Comment by DanielNoord** [comment]
Well, then I'm going to be `-1` on this proposal (although a smaller proposal I could perhaps agree with).

Overwriting methods of parent functions is perfectly valid and shouldn't raise a linter message.

```python
class Food:
   def type(self):
      print('I'm food')

class Vegetable(Food):
   def type(self):
      print('I'm a vegetable')

class Carro(Vegetable):
   def type(self):
      print('I'm a carrot')
```

Overwriting like this can help narrow down the return value in child classes.

**Comment by socketpair** [comment]
@DanielNoord Yes, your last example MUST NOT raise warnings. When I told about __MRO__  I was wrong. We should consider only parent classes, not all in MRO.

**Comment by jacobtylerwalls** [comment]
>  in real life, in most cases as I think, not calling base method is an error, again, in cases when the base method is normal.

Thanks for the proposal, but I find this likely to be viewed as extremely opinionated. -1

**Comment by socketpair** [comment]
Okay, if you think so, can we make this check in a way that does not break standard CI systems (low severity or so)? I will set this check as important in our CI.

Also, I strongly believe that redefining method without calling original (where original has some code) is rare. Possibly you have some way to check a new version of Pylint on a large codebase before merging ? 

P.S.
This case was really annoying bug that hard to find. That's why I want the check to be in Pylint.

**Comment by socketpair** [comment]
Actually, there is a mix of problems:
1. Redefining a method without calling super(). This is a case you concern on.
2. Redefining a method without calling a super in cases where both parents have this method. Actually my case.

The second case is VERY rare. And also, having same method in both parents is a bug as I think, unless both of them called explicitly.

And third case appear:
3. If a class has two same-named methods inherited from base classes hierarchy, and it does not redefine the method somehow, then it's a bug. Technically is not (because result depends on order of base classes in class definition), but really it's a bug (that's why linters exist).

**Comment by jacobtylerwalls** [comment]
> on the class definition line (i.e. the parentage is incorrect and should be fixed) 

I think it's hard to predict what should be "fixed", that's the nature of refactoring, but the advantage of putting it on the class line is to point users to exactly why pylint complained -- it's that the resulting behavior when using the class is fragile against the specific order you happened to use, and you should opt-in by silencing a warning if you want to accept the consequences.

**Comment by socketpair** [comment]
@jacobtylerwalls 

> Warn if a class has multiple direct parents that each define a method name besides __init__

Yes, but not precise. Warning should be issued only if the class does not reimplement the method. Also, using `super().method` in reimplementation should not be OK in this case (another warning).

**Comment by Pierre-Sassoulas** [comment]
> Possibly you have some way to check a new version of Pylint on a large codebase before merging ?

We have primer to show the result of a change on a selected number of repository.

**Comment by DanielNoord** [comment]
> 1. If parent classes have the same-named methods, reimplementation is required. "have" means not only own, but also in  class hierarchy. Also, different member types (like method and field) are not allowed to mix.

I think there is some agreement on that this could be a refactoring message. To give this issue some traction let's discuss a potential name. I propose to call it `overlapping-method-from-parents`, but I'm not 100% sold.
The warning should obviously include the name of the overlapping methods.

> 2. it's not allowed to use `super().same_method` in reimplementation in case number 1.

Both @jacobtylerwalls and I seem to agree that this is too opinionated for `pylint` to maintain. The reach of `pylint` is too large and the time we have to maintain it is too little to support all different opinionated checks that we could think of. 
See also https://github.com/PyCQA/pylint/pull/6901 which was a valid suggestion but is just a little too opinionated to add the associated maintenance burden in this repository.

You're obviously more than welcome to create a `pylint` extension yourself and I'd even be able to provide some assistance in doing so, but please continue that discussion outside of this issue/repo.

**Comment by jacobtylerwalls** [comment]
No worries, I'm finding the discussion helpful. My objection was mostly to warning about reimplementing parent methods without super _in general_, but that doesn't seem to be what we're talking about anymore. What do you think of the following?

**Overlapping concrete method names between direct parents (not between a parent and grandparent):**
  - not reimplemented on child: warn on child class definition
  - reimplemented on child: no warning
  - reimplemented on child with super(): warn on child class method (this is mutually exclusive with the earlier warning)

Overlapping isn't intuitive to me at least exactly what it refers to (or what's being guarded against.) It could be called `order-dependent-parent-method`? `order-dependent-method-resolution`?

@DanielNoord, does that last bullet strike you as too aggressive? I think @socketpair has a point that bullet points 3 and 1 basically point to the same thing: you're relying on order-dependence in a way that is easy to miss when the number of bases starts to grow.

**Comment by DanielNoord** [comment]
> `order-dependent-method-resolution`

I like this one with a clear and informative  message warning.

> @DanielNoord, does that last bullet strike you as too aggressive? I think @socketpair has a point that bullet points 3 and 1 basically point to the same thing: you're relying on order-dependence in a way that is easy to miss when the number of bases starts to grow.

I still feel like there are legitimate use cases where people know which method will be looked up. On the other hand, that's partially what a linter is for: telling you about potentially dangerous patterns.
I also don't know how I feel about raising the same message on class definitions and `super()` calls. I'm not sure if there is precedent for this, but it might be confusing?

Perhaps `order-dependent-resolution` (so we can also raise on property and attributes) and `order-dependent-super-resolution`?

**Comment by socketpair** [comment]
@jacobtylerwalls "between parents". The following case must trigger a message:

```python
class A:
    def close(self):
        pass

class B(A):
    pass

class C:
    def close(self):
        pass

class Bug(B, C):
    pass
```

@DanielNoord 

> I also don't know how I feel about raising the same message

I think it should be two different messages.

> Perhaps order-dependent-resolution (so we can also raise on property and attributes) and order-dependent-super-resolution?

I like these problem codes.

Regarding other information in previous two messages, I totally agree.

**Comment by jacobtylerwalls** [comment]
Assigning myself as I already started fiddling with adding this in the `MisdesignChecker`.

**Comment by jacobtylerwalls** [comment]
Thanks @NeilGirdhar you're not intruding at all, more input is welcome.

> Same as above. There's nothing wrong or even out of the ordinary about reimplmenting a method and calling super. That's exactly what you do with __init__ and __enter__, and other such method.

I agree with you here, so I just want to clarify. (Looking back, I was speaking in shorthand, sorry!) See:

> My objection was mostly to warning about reimplementing parent methods without super in general, but that doesn't seem to be what we're talking about anymore.

Instead, in my later comment, I was thinking of "reimplementing without super()" potentially being worth a warning _if_ we have already determined that two direct parents ("left to right" in an mro, not "up and down") had non-identical implementations. (Forgive relying on formatting to convey meaning, it was all supposed to be indented under the condition in bold.) That should mean your augmenting examples are not in jeopardy here, yes?

It's the `Child3` case here where the two parents have nonidentical implementations of `__repr__`:
https://github.com/PyCQA/pylint/blob/b342b01a4647c81f352e27e007b0d73a73ff2a3d/tests/functional/o/order_dependent_resolution.py#L36-L39


Given that, do you have opinions about #7007? I welcome more input, and as I said there, I wanted to open a WIP PR to get more data from the primer bot.
***
> This is extremely flimsy, in my opinion:

Yes, that's an argument for updating the `good.py` examples in #7007.

**Comment by NeilGirdhar** [comment]
> Instead, in my later comment, I was thinking of "reimplementing without super()" potentially being worth a warning if we have already determined that two direct parents ("left to right" in an mro, not "up and down") had non-identical implementations.

First of all, I think the only time that two parents should have implementations is when one or both call super.  If only one calls super, then it should be a child of the other.  Otherwise, I think it's a design error.

If you are "reimplementing without calling super", then you're overriding.  In an ideal world, I think overriding should have a decorator or comment to say that you know you're overriding behavior.  So I agree with you about emitting a warning.  Too bad there's no decorator in `typing` to declare that you're overriding and that you don't want any warning for this.

> That should mean your augmenting examples are not in jeopardy here, yes?

Right, since the error you're proposing is for when you don't call super, then not only does it not put the augmenting examples in jeopardy, but it actually helps people do good augmenting by warning them when they forgot to call super, which is a common mistake!

**Comment by NeilGirdhar** [comment]
> I agree, and the good news is that if you do adopt that later pattern you demonstrated:

I think that's too late to spit an error out.  I think class `C` should not be okay.  I don't think there's a good reason to do `A.close(self)`.  If you want to call super, call super.  If you don't have a way to make `A` call super because it's in library code that you can't change, then pylint should still create an error, but you can ignore the error.

> Right. Reordering won't help. But the fact that it depends on order at all is a symptom of the problem tractable for a linter to identify. 

Right.  I like your `Child1` failure.

**Comment by NeilGirdhar** [comment]
Now that I think about it some more, I'm not sure about the name "order-dependent resolution".  After all, there's nothing wrong with base class ordering making a difference.  The issue is that in `Child1` and `Child3`, `Parent2.__repr__` is _implicitly hidden_. In `Child2`, it's okay because it's _explicitly hidden_ due to the override.

Do you think it makes sense to add this example to your tests?
```python
class HasClose:
    def close(self):
        pass

class A(HasClose):
   def close(self):
      print('A')
      super().close()

class B(HasClose):
   def close(self):
      print('B')
      super().close()

class C(A, B):
   def close(self):
      print('C')
      super().close()
```
Here, nothing is hidden.

**Comment by jacobtylerwalls** [comment]
Thanks, those are good suggestions on names and tests.

**Comment by jacobtylerwalls** [comment]
@NeilGirdhar it sounds like you strongly disagree with OP's example, then, if you expect no message for `class C`. I think there's enough disagreement here that we should iterate slowly and carefully: I'll reduce the scope of #7007 to just the examples without super so that your test case passes.

For future implementers, this is what I was working with, at the end of the new method in #7007:
<details>

        for parent_meth in methods_defined_in_multiple_bases:
            if parent_meth in missing_reimplementations:
                continue
            this_meth = next(node.ilookup(parent_meth.name), None)
            if this_meth is None:  # pragma: no cover
                continue
            meth_frame = this_meth.frame()
            for call_node in this_meth.nodes_of_class(nodes.Call):
                if call_node.frame() is not meth_frame:
                    continue
                if isinstance(call_node.func, nodes.Attribute):
                    if not isinstance(call_node.func.expr, astroid.objects.Super):
                        continue
                elif isinstance(call_node.func, nodes.Name):
                    if call_node.func.name != "super":
                        continue
                else:
                    continue
                self.add_message(
                    "hidden-super-method",
                    node=call_node,
                    confidence=INFERENCE,
                    args=(
                        this_meth.name,
                        ", ".join(c.name for c in parents_to_mention_in_warning),
                    ),
                )
```

**Comment by jacobtylerwalls** [comment]
More notes for future -- IIUC the reason @NeilGirdhar doesn't object to `class C` but accepts a warning for `Child3` in https://github.com/PyCQA/pylint/issues/6992#issuecomment-1163850865 is because he is thinking of an all-paths-delegate-to-super concept, which I haven't taken up at the moment.

**Comment by NeilGirdhar** [comment]
@jacobtylerwalls  Thank you very much.

> is because he is thinking of an all-paths-delegate-to-super concept, which I haven't taken up at the moment.

FYI this is the pattern that methods like `__init__`, `__enter__`, `__exit__`, and `numpy.ndarray.__array_finalize__` already implement.

The pattern where superclass methods are explicitly called (`A.close(self)`) is imho an anti-pattern.

**Comment by bje-** [comment]
I have a related issue. I have a pair of classes that both define a method called `reset()`. They both make sense in the context of their own classes. I am using multiple inheritence to bring these two together and recently encountered a bug because the derived class `reset()` forgot to call both parent `reset()` methods. Simplified example appears below:

```
class Parent1():
  def reset(self):
    pass

class Parent2():
  def reset(self):
    pass

class Child(Parent1, Parent2):
  def reset(self):
    Parent1.reset(self)
    # Parent2.reset isn't called -- whoops.
```

I wish Pylint had caught this one for me.
